### ISSUE TITLE
Django's deconstructible decorator causes to lose type information

### ISSUE THREAD

**Issue by karolyi** [microsoft/pyright#3840]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Hey,

I'm using `djmoney.money.Money` (django-moneyed) to use `Money` classes in my django project. That class has a `@deconstructible` decorator in front of it. Using the class, pyright loses the typing information for some reason. All I could figure out is that it's the `deconstructible` decorator that causes it.

The source of the decorator is:
```python
def deconstructible(*args, path=None):
    """
    Class decorator that allows the decorated class to be serialized
    by the migrations subsystem.

    The `path` kwarg specifies the import path.
    """
    def decorator(klass):
        def __new__(cls, *args, **kwargs):
            # We capture the arguments to make returning them trivial
            obj = super(klass, cls).__new__(cls)
            obj._constructor_args = (args, kwargs)
            return obj

        def deconstruct(obj):
            """
            Return a 3-tuple of class import path, positional arguments,
            and keyword arguments.
            """
            # Fallback version
            if path:
                module_name, _, name = path.rpartition('.')
            else:
                module_name = obj.__module__
                name = obj.__class__.__name__
            # Make sure it's actually there and not an inner class
            module = import_module(module_name)
            if not hasattr(module, name):
                raise ValueError(
                    "Could not find object %s in %s.\n"
                    "Please note that you cannot serialize things like inner "
                    "classes. Please move the object into the main module "
                    "body to use migrations.\n"
                    "For more information, see "
                    "https://docs.djangoproject.com/en/%s/topics/migrations/#serializing-values"
                    % (name, module_name, get_docs_version()))
            return (
                path or '%s.%s' % (obj.__class__.__module__, name),
                obj._constructor_args[0],
                obj._constructor_args[1],
            )

        klass.__new__ = staticmethod(__new__)
        klass.deconstruct = deconstruct

        return klass

    if not args:
        return decorator
    return decorator(*args)
```

**To Reproduce**
Probably any class decorated with this decorator will lose its typing information with pyright.

**Expected behavior**
They should keep their types?

**Screenshots or Code**
A quick test I've put together with using `DefaultDict` and `Money`:
```python
from collections import defaultdict
from typing import DefaultDict
from moneyed import EUR
from djmoney.money import Money

MONEY_ZERO = Money(0, EUR)

a: DefaultDict[int, Money] = defaultdict(lambda: MONEY_ZERO)

x = a[5]
```
Screenshots of what pyright thinks about types in this code:
![image](https://user-images.githubusercontent.com/987055/185599789-ec3f5036-7da5-494f-bfd5-02ea6b22925a.png)
![image](https://user-images.githubusercontent.com/987055/185599896-648aba96-5966-41a1-a4aa-da3cbb9f0b4d.png)

Maybe you could enlighten me as to what happens here that leads to losing typing information?

**Comment by erictraut** [comment]
Which packages (and versions) would I need to install to repro the issue? Are these packages typed? (Based on the code you pasted above, it appears that it's not typed.) If not, do you have type stubs for them? Type checking relies on accurate type information either inlined within the library or in the form of type stubs.

**Comment by karolyi** [comment]
I'm using django2.2, but 3.2 seems to have the same function definition for the decorator.

Django stubs is installed in the project where I use the function, and it defines the constructor as follows:
```python
def deconstructible(*args: Any, path: Optional[Any] = ...) -> Any: ...
```

Hence the `Any` type, I guess.

When installing django==3.2 and django-money==3.0.0 without the stubs, the return type becomes `Unknown`.

Does this help?

**Comment by erictraut** [comment]
The problem is in `django-stubs`. The file `django-stubs/utils/deconstruct.pyi` declares the `deconstructable` function as follows:

```python
def deconstructible(*args: Any, path: Optional[Any] = ...) -> Any: ...
```

It should probably be declared something like this instead:

```python
_T = TypeVar("_T")
_TCallable = TypeVar("_TCallable", bound=Callable[Any, ...])

@overload
def deconstructible(_type: type[_T]) -> type[_T]: ...
@overload
def deconstructible(
    *, path: Optional[Any] = ...
) -> Callable[[_TCallable], _TCallable]: ...
```

**Comment by karolyi** [comment]
Thanks. I will make a PR for django-stubs.

**Comment by karolyi** [comment]
I've just added a merge request to django-stubs, slightly modifying your code (probably a typo).

Thanks for the help!

**Comment by erictraut** [comment]
Happy to help. Thanks for submitting a PR. That will help other django users avoid this problem in the future.
